src/librustc/mir/transform.rs

```diff
@@ -8,9 +8,79 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use mir::repr::Mir;
+use mir::repr::{Mir, BasicBlockData, BasicBlock};
+use mir::mir_map::MirMap;
 use middle::ty::ctxt;
 
-pub trait MirPass {
-    fn run_on_mir<'tcx>(&mut self, mir: &mut Mir<'tcx>, tcx: &ctxt<'tcx>);
+/// Contains various metadata about the pass.
+pub trait Pass {
+    /// Ordering of the pass. Lower value runs the pass earlier.
+    fn priority(&self) -> usize;
+    // Possibly also `fn name()` and `fn should_run(Session)` etc.
+}
+
+/// Pass which inspects the whole MirMap.
+pub trait MirMapPass: Pass {
+    fn run_pass<'tcx>(&mut self, tcx: &ctxt<'tcx>, map: &mut MirMap<'tcx>);
+}
+
+/// Pass which only inspects MIR of distinct functions.
+pub trait MirPass: Pass {
+    fn run_pass<'tcx>(&mut self, tcx: &ctxt<'tcx>, mir: &mut Mir<'tcx>);
+}
+
+/// Pass which only inspects basic blocks in MIR.
+///
+/// Invariant: The blocks are considered to be fully self-contained for the purposes of this pass –
```

Comment:
Why do we need this invariant? For the `MirPass` implementor it's irrelevant, as it's the same as making sure that the visitor and the `run_pass` method don't modify the `Mir` object. For the compiler there's no advantage to knowing this as I see it.